Add function to export user report data from Gophish

JCantu248 opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

Function was added to gather user report data for all campaigns associated with a given assessment. The data is obtained by using the GoPhish API, and written to a JSON file for later use by PCA operators. Each campaign is written to its own JSON file. 

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

As part of the PCA team's effort to develop Live PCA, it is requested that user report data (that is targeted user clicking on links) be obtained from GoPhish and collected into a format that allows for import into the PCA database. This will allow for archival storage of assessment and campaign data. 


## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

Code base was tested in the Staging COOL environment. Two new campaigns were set up in GoPhish running in Staging COOL, and after their completion, the GoPhish Export command was run with the appropriate assessment name.  A JSON for each campaign was generated, with properly formatted JSON. 

Unfortunately since we don't have corresponding assessment and campaign documents from GoPhish loaded into the PCA DB, I wasnt able to test importing the generating JSON. What I did instead was wrote the first_report datetime into a known good and valid user report JSON file, and forced an overwrite of the document. I did this to make sure that the datetime format I was using for first_report was matching what the importer and PCA DB were expecting. 


## 📷 Screenshots (if appropriate) ##

<!-- Remove this section and header if not needed -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

JCantu248 committed:
Add export_user_reports to gophish_export.py

Adding a function to build an export user_report JSON files for each campaign in an assessment.

JCantu248 committed:
Minor code change to new function

dav3r committed:
Bump tools version from 0.0.5 to 0.0.6

dav3r committed:
Bump project version from 0.0.4 to 0.0.5

dav3r committed:
Bump project version from 0.0.5 to 0.0.5-rc.1

JCantu248 committed:
bug fix related to str and NoneType comparisons

JCantu248 committed:
Merge branch 'export_user_reports' of github.com:cisagov/gophish-tools into export_user_reports
Testing build of new export functions

JCantu248 committed:
Bump project version from 0.0.5-rc.1 to 0.0.5-rc.2

JCantu248 committed:
Fixing bug by adding string to datetime parser

JCantu248 committed:
Fixing datetime parser

JCantu248 committed:
Still fixing datetime bugs

JCantu248 committed:
Bump project version from 0.0.5-rc.4 to 0.0.5-rc.5

JCantu248 committed:
Bump project version from 0.0.5-rc.5 to 0.0.5-rc.6

JCantu248 committed:
Bump project version from 0.0.5-rc.6 to 0.0.5-rc.7

JCantu248 committed:
Merge branch 'develop' of github.com:cisagov/gophish-tools into export_user_reports
Merging develop into feature branch

JCantu248 committed:
Bump project version from 0.0.5-rc.7 to 0.1.0

JCantu248 committed:
Bump project version from 0.1.0 to 0.1.0-rc.1

JCantu248 committed:
WIP. Fixing datetime compatibility with the pca-import tool

JCantu248 committed:
Bump project version from 0.1.0-rc.2 to 0.2.0

JCantu248 committed:
Updating README.md to reflect changes

Added User Report Field Dictionary to describe the user_report JSON files that will be generated by gophish_export

dav3r commented on README.md:
````diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
 
+## User Report Field Dictionary ##
+
+User report JSONs are also exported by ```gophish-export```. User reports
````
Single backticks are sufficient here:
```suggestion
User report JSONs are also exported by `gophish-export`. User reports
```

dav3r commented on README.md:
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
+| assessment | A corresponding assessment ID that this user report is associated with. | string | | yes |
+| campaign | A corresponding campaign ID that this user report is associated with. | string| | yes |
+| first_report | The first report (click) generated by a targeted user. Formatted by YYYY-MM-DDT: hh:mm.ss"| datetime | | yes |
+| total_num_reports | The total number of user reports received during a campaign.| integer | | no
```
I recommend sorting these rows alphabetically by field name.

JCantu248 replied:
I put them in this order because this is the order I have them on in the JSON file. Are you sure you wanna change the field order?

jsf9k replied:
The order of the fields in the JSON file shouldn't be relevant, right?  It isn't a CSV.  I think @dav3r is right here.

If part of your code depends on the order of JSON dictionary keys that probably indicates a problem, since JSON does not make any promises about such ordering.  If you parse a JSON and then re-export it the order of the keys can change depending on the hash map implementation being used internally.

JCantu248 replied:
See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

dav3r replied:
@JCantu248, https://github.com/cisagov/gophish-tools/pull/33/commits/bac2cb88e00b183bdbae8f990c3ad5a48b01ef05 does not correct the alphabetical order.

Also, please use more descriptive messages in your commits: `Merging remote` is not particularly helpful here.  We strongly encourage everyone to follow the commit message style and guidelines described in [this post](https://chris.beams.io/posts/git-commit/#seven-rules), as referenced in our [Development Guide](https://github.com/cisagov/development-guide#git-github-and-you).

dav3r replied:
Thanks @JCantu248 for fixing the alphabetization in https://github.com/cisagov/gophish-tools/pull/33/commits/cc24b7780b4ff24ca8f43633d2dce6a214d86892. 👍

dav3r commented on README.md:
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
+| assessment | A corresponding assessment ID that this user report is associated with. | string | | yes |
+| campaign | A corresponding campaign ID that this user report is associated with. | string| | yes |
+| first_report | The first report (click) generated by a targeted user. Formatted by YYYY-MM-DDT: hh:mm.ss"| datetime | | yes |
+| total_num_reports | The total number of user reports received during a campaign.| integer | | no
```
Clean up wording a bit and remove PCA-specific language:
```suggestion
| Name | Description | Type | Default | Required |
|------|-------------|:----:|:-------:|:--------:|
| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as `null`. | string | `null` | no |
| assessment | Assessment ID that this user report is associated with. | string | | yes |
| campaign | Campaign ID that this user report is associated with. | string | | yes |
| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDT: hh:mm.ss" | datetime | | yes |
| total_num_reports | Total number of user reports received during a campaign. | integer | | no |
```

dav3r commented on README.md:
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
+| customer | A corresponding customer ID that this user report document is associated with. It is left null as it will be filled in later by a PCA operator or another tool prior to importation | string | null | No |
+| assessment | A corresponding assessment ID that this user report is associated with. | string | | yes |
+| campaign | A corresponding campaign ID that this user report is associated with. | string| | yes |
+| first_report | The first report (click) generated by a targeted user. Formatted by YYYY-MM-DDT: hh:mm.ss"| datetime | | yes |
```
Is `YYYY-MM-DDT: hh:mm.ss` really the correct format or should it be `YYYY-MM-DDThh:mm.ss`?

JCantu248 replied:
See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

dav3r replied:
It looks like you corrected the format in https://github.com/cisagov/gophish-tools/pull/33/commits/ccdf910c06d009dca6304f8223285cb2d9d99042, but then overwrote it with https://github.com/cisagov/gophish-tools/commit/bac2cb88e00b183bdbae8f990c3ad5a48b01ef05.  I assume that you did not mean to do that.  Please let me know when it's fixed.

JCantu248 replied:
Fixed here 71d9fc125456322326ea57a91ce7d86fbf07d45e

dav3r replied:
Thanks!

dav3r commented on src/_version.py:
```diff
@@ -1,2 +1,2 @@
 """This file defines the version of this project."""
-__version__ = "0.0.4"
+__version__ = "0.2.0"
```
What happened to `0.1.0`?

JCantu248 replied:
Not sure how I messed up the bump_version. I manually fixed it here: 3cd389cce6498bf6579e9ce3c47fdb0968e9febf

dav3r commented on README.md:
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
+| assessment | A corresponding assessment ID that this user report is associated with. | string | | yes |
+| campaign | A corresponding campaign ID that this user report is associated with. | string| | yes |
+| first_report | The first report (click) generated by a targeted user. Formatted by YYYY-MM-DDT: hh:mm.ss"| datetime | | yes |
+| total_num_reports | The total number of user reports received during a campaign.| integer | | no
```
I thought over in [this discussion](https://github.com/cisagov/pca-core/pull/5#discussion_r651079451), we were saying that `total_num_reports` cannot be blank, i.e. it is required?

JCantu248 replied:
See bac2cb88e00b183bdbae8f990c3ad5a48b01ef05

dav3r replied:
It looks like you changed this field to be required in ccdf910, but then overwrote it with bac2cb8. I assume that you did not mean to do that. Please let me know when it's fixed.

JCantu248 replied:
See cc24b7780b4ff24ca8f43633d2dce6a214d86892

dav3r replied:
https://github.com/cisagov/gophish-tools/commit/cc24b7780b4ff24ca8f43633d2dce6a214d86892 fixes the alphabetization issue, but `total_num_reports` is still listed as optional.  Did you mean to put this comment in the conversation above?

JCantu248 replied:
Nah, it was an oversight. See 51786dc6d5b6d82de55ecbf822644fe2fe43ce71

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,44 @@ def get_application(rawEvent):
+
+        # iterate over clicks and find the earliest click
+        for click in campaign["clicks"]:
+            if (
+                first_report is None
+                or datetime.strptime(click["time"].split(".")[0], "%Y-%m-%dT%H:%M:%S")
+                < first_report
+            ):
+                first_report = datetime.strptime(
+                    click["time"].split(".")[0], "%Y-%m-%dT%H:%M:%S"
+                )
```
Here's a slightly cleaner way to write this:
```suggestion
            click_time = datetime.strptime(
                click["time"].split(".")[0], "%Y-%m-%dT%H:%M:%S"
            )
            if first_report is None or click_time < first_report:
                first_report = click_time
```

JCantu248 replied:
See 724b35ac39da24c61713cc0dd0addb7a5f3befe5

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,44 @@ def get_application(rawEvent):
+
+        user_report_doc["customer"] = None
+        user_report_doc["assessment"] = assessment_id
+        user_report_doc["campaign"] = campaign["id"]
```
Is there any difference between `campaign_id` and `campaign["id"]` here?  If not, I'd simply go with:
```suggestion
        user_report_doc["campaign"] = campaign_id
```

JCantu248 replied:
I tried that before during my testing. The campaign_id that iterates over campaign_ids, which is in turn returned from get_campaign_ids() is an integer value. campaign["id"] is a string value. The schema for the campaign field in the user_report is defined as a string as well, as thats what campaign _ids were defined as throughout the PCA database and our legacy data. Im using the campaign["id"] so I can load string data into the current schema.

dav3r replied:
Thanks for the explanation.  In the interest of more readable code, how about something like this?
```
# get_campaign_ids() returns integers, but user_report_doc["campaign"]
# expects a string
user_report_doc["campaign"] = str(campaign_id)
```

JCantu248 replied:
See 724b35ac39da24c61713cc0dd0addb7a5f3befe5

dav3r commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,44 @@ def get_application(rawEvent):
+        ).stats.clicked
+
+        logging.info(
+            f"Writing out user reports for all campaigns in assessment: {assessment_id}"
```
Since this logging message is inside the `for campaign_id in campaign_ids` loop, it should indicate which campaign is being output:
```suggestion
            f"Writing out user report for campaign {campaign_id} in assessment {assessment_id}"
```

dav3r reviewed (changes requested):
I have some questions, comments, and feedback- please let me know if you have any questions about what I wrote.

JCantu248 committed:
Update README.md

Change to single backticks

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

JCantu248 committed:
Update README.md

Cleaning up wording on User Report field descriptions

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

JCantu248 committed:
Fixes to README.md

Changes to user_report field order and datetime format

JCantu248 committed:
Merging remote

JCantu248 committed:
Fixing version number

BenBreaksThings commented:
Chiming in out of turn here, but many of the items above are characterized as recommendations or are instances of refactoring. Not that the code is broken, just preferential construct. Can we adopt a more ops focused mindset here and accept that the working code that isn't doing anything dangerous need not be prevented from progressing due to JSON alphabetizing?

JCantu248 committed:
Update src/tools/gophish_export.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

felddy commented:
> Chiming in out of turn here, but many of the items above are characterized as recommendations or are instances of refactoring. Not that the code is broken, just preferential construct. Can we adopt a more ops focused mindset here and accept that the working code that isn't doing anything dangerous need not be prevented from progressing due to JSON alphabetizing?

It's all good.  All chiming is welcome, at all times.

We do indeed have preferential constructs, and they're usually not just whims.   We strive to give our codebase as much longevity as possible.  This means that we hope that it will continue to be in useful long after we're done writing it.  The more readable, and uniform we can make our projects, the better the chance they will survive changes in maintainers and owners.

It's ok to disagree with a suggestion but we'd ask that you provide your reasoning when you do.  Our coding best-practices are always evolving and we're always looking to improve them.  

I don't think any of the conversations and suggestions above are out of the ordinary.  This is actually less of a barrage than I usually get on my PRs.   😅 

In any case, thank you for contributing and running the gauntlet with us.  It's not always fun, but the project will be better for it.

JCantu248 committed:
Code changes to logic for getting the first_report, and changes to campaign_id

JCantu248 committed:
Merge branch 'export_user_reports' of github.com:cisagov/gophish-tools into export_user_reports

Merging changes from remote into local branch.

JCantu248 committed:
Restoring fixes to README.md

JCantu248 committed:
Fixing total_num_reports. Forgot to change required from no to yes.

JCantu248 committed:
Fixing datetime format overwrite to the correct format

dav3r commented on README.md:
```diff
@@ -21,7 +21,8 @@ campaign assessors to simplify the process of managing GoPhish campaigns.
 * `gophish-export` - Exports all the data from an assessment within GoPhish
-  into a single JSON file.
+  into a single JSON file. In addition, user report JSONs for each campaign in
+  an assessment will also be generated, for later import into the PCA database.
```
Remove reference to PCA- this repo should stand on it's own as a Gophish helper utility, regardless of whether or not it is used for PCAs.
```suggestion
  an assessment will also be exported.
```

JCantu248 replied:
See 2d96175613954e3c341aa38614d80369113217af

dav3r replied:
Thanks!  FYI, you can use [GitHub's "Commit suggestion" feature](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) to save time when making changes like this in the future.

dav3r commented on README.md:
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
+
+User report JSONs are also exported by `gophish-export`. User reports
+summarize data from targeted user clicks on phishing emails generated by a
+campaign. These JSON files are later imported into the PCA database.
```
Remove reference to PCA- this repo should stand on it's own as a Gophish helper utility, regardless of whether or not it is used for PCAs.
```suggestion
campaign.
```

JCantu248 replied:
See 2d96175613954e3c341aa38614d80369113217af

JCantu248 committed:
Removing references to PCA from project documentation

dav3r reviewed (approved):
Thanks for working through all of my feedback - this looks great now! 🎸

dav3r commented:
@JCantu248 - I just noticed that you still need to pull in the latest changes from `develop`.  Please take care of that when you can.

JCantu248 committed:
Merge branch 'develop' of github.com:cisagov/gophish-tools into export_user_reports
Merging latest changes from develop into export_user_repo prior to closing PR #33

mcdonnnj commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,43 @@ def get_application(rawEvent):
+            campaign_id=campaign_id
+        ).stats.clicked
+
+        logging.info(
+            f"Writing out user report for campaign {campaign_id} in assessment {assessment_id}"
+        )
```
Logging calls should use the `printf` style notation, but I know there are a lot of them in this repository. Please make an issue in the repo to switch all `logging` library calls to remove f-strings and use `printf` style formatting.

dav3r replied:
For a good explanation of why we no longer put f-strings in logging statements, check out [this post](https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/).

JCantu248 replied:
See [here for issue](https://github.com/cisagov/gophish-tools/issues/34)

mcdonnnj commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,43 @@ def get_application(rawEvent):
 
 
+def export_user_reports(api, assessment_id):
+    """Build and export a user_report JSON object for each campaign in an assessment."""
```
```suggestion
    """Build and export a user_report JSON file for each campaign in an assessment."""
```

JCantu248 replied:
Word changed in 3c54a97003c86809fdefc23a46a7ca562492e01c

mcdonnnj commented on src/tools/gophish_export.py:
```diff
@@ -253,6 +253,43 @@ def get_application(rawEvent):
+            if first_report is None or click_time < first_report:
+                first_report = click_time
+
+        user_report_doc["customer"] = None
```
Is a `null` the desired output in the JSON for this field? Your README change indicates the field is a `string`. A `null` and a `string` are not the same, so I am suggesting we instead output an empty string for this field.

I have also added a comment to provide some insight on why such a field is in the output in the first place. @dav3r I would appreciate your input on making this comment better.

```suggestion
        # The "customer" field is a placeholder added for operator convenience when
        # working with the JSON file created.
        user_report_doc["customer"] = ""
```

dav3r replied:
@mcdonnnj - I like the comment.  I'd only make a minor change to minimize any confusion:
```
# The "customer" field is a placeholder added for operator convenience when
# working with the JSON file created.
```

jsf9k replied:
This is a good point to clarify.  Thanks for pointing this out @mcdonnnj.

JCantu248 replied:
This is changed in 3c54a97003c86809fdefc23a46a7ca562492e01c

mcdonnnj commented on README.md:
```diff
@@ -168,6 +169,20 @@ with the command `pca-wizard-templates  --emails`.
+| campaign | Campaign ID that this user report is associated with. | string | | yes |
+| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as `null`. | string | `null` | no |
+| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDThh:mm.ss" | datetime | | yes |
+| total_num_reports | Total number of user reports received during a campaign. | integer | | yes |
```
`Default`/`Required` doesn't make much sense for a JSON output. I also updated one of the descriptions to reflect a change I suggested in the code.

```suggestion
| Name | Description | Type |
|------|-------------|:----:|
| assessment | Assessment ID that this user report is associated with. | string |
| campaign | Campaign ID that this user report is associated with. | string |
| customer | Customer ID that this user report document is associated with. Note that Gophish does not contain this information, so `gophish-export` will always export it as an empty string. | string |
| first_report | First report (click) generated by a targeted user. Format: "YYYY-MM-DDThh:mm.ss" | datetime |
| total_num_reports | Total number of user reports received during a campaign. | integer |
```

JCantu248 replied:
Changes incorporated in cf4906f0b51ec10f488d8d013bed0d28003dd4ce

mcdonnnj commented on README.md:
```diff
@@ -21,7 +21,8 @@ campaign assessors to simplify the process of managing GoPhish campaigns.
 * `gophish-complete` - Completes a campaign in GoPhish and/or outputs a
   GoPhish campaign summary.
 * `gophish-export` - Exports all the data from an assessment within GoPhish
```
Please make an issue to replace all instances of `GoPhish` with `Gophish` in this project. We've slowly been making the change in other projects as we do updates and I don't want to scope creep this PR.

JCantu248 replied:
See [here](https://github.com/cisagov/gophish-tools/issues/35) for the new issue

mcdonnnj reviewed (changes requested):
I had some suggestions and a couple of issues I would like created. Thanks for your work!

jsf9k reviewed (approved):
:+1:

JCantu248 committed:
Responding to feedback. Changes to docstring and null type to empty string.

Minor rewording of the docstring for export_user_reports and changing the user_report_doc to default the customer field from null to the empty string

JCantu248 committed:
Incorporating feedback change requests for README.md

mcdonnnj reviewed (approved):
I believe you have addressed all of my concerns. Thanks for working with us to get this finished!
